src/main/scala/util/DelayQueue.scala

```diff
@@ -0,0 +1,40 @@
+// See LICENSE.SiFive for license details.
+
+package freechips.rocketchip.util
+
+import chisel3._
+import chisel3.util._
+
+object DelayQueue {
+  def apply[T <: Data](sink: DecoupledIO[T], source: DecoupledIO[T], timer: UInt, delay: UInt, depth: Int): Unit = {
+      val q = Module(new Queue(new Bundle {
+        val data = source.bits.cloneType
+        val time = UInt(timer.getWidth.W)
+      }, depth, flow=true))
+
+      q.io.enq.bits.data := source.bits
+      q.io.enq.bits.time := timer
+      q.io.enq.valid := source.fire
+      source.ready := q.io.enq.ready
+
+      sink.bits := q.io.deq.bits.data
+      sink.valid := q.io.deq.valid && ( (timer - q.io.deq.bits.time) >= delay)
+      q.io.deq.ready := sink.fire
+  }
+}
+
+object FixedDelayQueue {
+  def apply[T <: Data](sink: DecoupledIO[T], source: DecoupledIO[T], delay: Int, maxLatency: Int = 4096): Unit = {
```

Comment:
This can become
```
  apply(sink, source, delay.U, delay.U, maxLatency)
```

Reply:
Sorry I don't follow? What do you mean?

Reply:
Have a single object `DelayQueue`, with method overloading of `apply`. Note the different type signatures, depending on if you want a fixed delay (pass a Int) or a variable delay (pass a UInt and a max).

For `depth`, you would normally want to set this to be equal to the `maxDelay` right?

Reply:
Ah. I see. I can do that.

Reply:
And yea I missed the `maxDelay` comment. Updated as well.